channels/routing.py

```diff
@@ -100,6 +100,15 @@ async def __call__(self, scope, receive, send):
         path = scope.get("path_remaining", scope.get("path", None))
         if path is None:
             raise ValueError("No 'path' key in connection scope, cannot route URLs")
+
+        if "path_remaining" not in scope:
+            # We are the outermost URLRouter, so handle root_path if present.
+            root_path = scope.get("root_path", "")
+            if root_path and not path.startswith(root_path):
+                # If root_path is present, path must start with it.
+                raise ValueError("No route found for path %r." % path)
```

Comment:
This should probably mention `root_path` as well, but I was not sure how best to phrase it. Suggestions welcome.